Documentation/Yarn-Spec.md

```diff
@@ -0,0 +1,636 @@
+Any other options lines with the same indentation is considered a new option line and closes the block for that option.
+
+These rules are repeated for each option line until a non-option line with the same, or less indentation than the base indentation is encountered which closes the block and the option statement entirely.
+
+Options can be nested inside options.
+Not every option line needs to have blocks.
+
+##### Tabs vs Space
+
+The choice to require either tabs or spaces over the other is unspecified.
+Tabs and spaces shouldn't be mixed.
+Should there be a need to convert between them the conversion rate must be the same at all points in the project.
+The rate of conversion between tabs to spaces, and spaces to tabs, is unspecified.
+If there is a need to choose one, tabs should be preferred due to their improved accessibility over spaces.
+
+## Expressions
+
+_Expressions_ are mathematical chains of variables, values, functions, expressions, and operators that produce a single value as output.
+
+Expressions are not a statement but are a component of various statements and must only be used as part of a statement, they cannot exist in isolation.
+This means if you do want to show the result of an expression it will have to be wrapped inside an interpolated line.
+For example a line that is just `$numberOfCoins + 1` is invalid, but `{$numberOfCoins + 1}` is valid.
+
+Expressions are mostly used to control the flow of the if statement.
+
+### Values
+
+A _Value_ is a single concrete form of one of the supported types.
+All expressions, subexpressions, variables and functions, must resolve down into a value before it can be used.
+Examples of values include `1`, `true`, `"General Kenobi"`.
+
+### Types
+
+Yarn supports the following types and these must be supported by an implementing program:
+
+- number
+- boolean
+- string
+
+The precision, storage, and form of the number internally is unspecified, however it must support decimals.
+As an example of this in C# the `Decimal`, `Complex`, and `float` formats are valid (though some make more sense than others) but `int` is not.
+Numbers in expressions can be written as either integers or as decimals, but decimal numbers must use the `.` symbol as the decimal seperator of the the number.
+
+Strings must be capable of holding utf-8 values as this is what the yarn language is written in, but the internals of this is unspecified provided all valid utf-8 characters are supported.
+Strings in expressions must be encapsulated between `"` and `"` symbols.
+
+Booleans must be capable of representing the boolean logic values of `true` and `false`, however the specific implementation is undefined.
+Booleans must not be exposed as `1` and `0` to expressions even if they are represented this way internally by the implementing program.
+Booleans in expressions must be written as `true` for true and `false` for false.
+
+Additional types supported are unspecified.
```

Comment:
Recommend that additional types are discouraged to increase interoperability between Yarn implementations